File: craft_project_invoicing/events/sales_order.py
import frappe

def validate(doc, method):
    if doc.enable_project_invoicing and (doc.advance_percentage or doc.on_delivery_percentage or doc.retention_percentage):
        try:
            total = float(doc.advance_percentage or 0) + \
                    float(doc.on_delivery_percentage or 0) + \
                    float(doc.retention_percentage or 0)
                    
            if total != 100:
                frappe.throw(
                    title='Error',
                    msg='Total of Advance, On Delivery and Retention should be <b>100%</b>'
                )
        except ValueError:
            frappe.throw(
                title='Error',
                msg='Advance, On Delivery, and Retention percentages must be numeric values.'
            )



from frappe.model.mapper import get_mapped_doc

@frappe.whitelist()
def create_proforma_invoice(source_name, target_doc=None):
    doc = get_mapped_doc(
        "Sales Order",                     
        source_name,                       
        {
            "Sales Order": {               
                "doctype": "Proforma Invoice",  
                "field_map": {                  
                    "name": "against_sales_order",      
                    "customer": "customer",     
                    "transaction_date": "transaction_date",
                    "delivery_date" : "delivery_date"
                }
            },
            "Sales Order Item": {           
                "doctype": "Proforma Invoice Item",
                "field_map": {               
                    "item_code": "item_code",
                    "qty": "qty",
                    "rate": "rate",
                    "amount": "amount",
                    "description": "description",
                    "name": "so_item_detail",
                    "delivery_date" : "delivery_date"
                    
                }
            }
        },
        target_doc                           
    )
    return doc

File: craft_project_invoicing/public/js/sales_order.js
frappe.ui.form.on('Sales Order', {
	custom_order_invoicing_type: function (frm) {
		if (frm.doc.custom_order_invoicing_type && frm.doc.custom_order_invoicing_type == "Project Invoicing") {
			frm.set_value("enable_project_invoicing", 1);
		}
		else if (frm.doc.custom_order_invoicing_type && frm.doc.custom_order_invoicing_type == "Normal"){
			frm.set_value("enable_project_invoicing", 0);
		}
	},
	advance_percentage: function (frm) {
		let advance = frm.doc.advance_percentage ? parseFloat(frm.doc.advance_percentage) : 0;
		let retention = frm.doc.retention_percentage ? parseFloat(frm.doc.retention_percentage) : 0;
		
		frm.set_value("on_delivery_percentage", (advance || retention) ? (100 - (advance + retention)) : 0);
	},
	
	retention_percentage: function (frm) {
		let advance = frm.doc.advance_percentage ? parseFloat(frm.doc.advance_percentage) : 0;
		let retention = frm.doc.retention_percentage ? parseFloat(frm.doc.retention_percentage) : 0;
		
		frm.set_value("on_delivery_percentage", (advance || retention) ? (100 - (advance + retention)) : 0);
	},

	onload: function (frm) {
		if (frm.doc.enable_project_invoicing && frm.doc.retention_billed === 0) {
			frm.add_custom_button(__('Sales Invoice'), () => frm.events.make_sales_invoice(frm), __('Create'));
		}
	},

	make_sales_invoice: function (frm) {
		frappe.model.open_mapped_doc({
			method: "erpnext.selling.doctype.sales_order.sales_order.make_sales_invoice",
			frm: frm
		});
	},
	refresh: function(frm) {
        frm.add_custom_button("Proforma Invoice", function() {
            frappe.call({
                method: "craft_project_invoicing.events.sales_order.create_proforma_invoice",
                args: {
                    source_name: frm.doc.name
                },
                callback: function(r) {
                    if (r.message) {
                        frappe.model.sync(r.message);
                        frappe.set_route("Form", r.message.doctype, r.message.name);
                    }
                }
            });
        }, "Create");
    }
});
